### PR TITLE
db_bench supports for generating random variable sized value.

### PR DESCRIPTION
**Pull request by burtonli**
1. `db_bench` now supports `value_size_distribution_type`, `value_size_min`, `value_size_max` options for generating random variable sized value. 
2. Added `blob_db_compression_type` option for BlobDB to enable blob compression.

### CODE REVIEW

File: tools/db_bench_tool.cc [hunk 1]
```diff
@@ -245,7 +245,15 @@ DEFINE_int32(threads, 1, "Number of concurrent threads to run.");
 DEFINE_int32(duration, 0, "Time in seconds for the random-ops tests to run."
              " When 0 then num & reads determine the test duration");
 
-DEFINE_int32(value_size, 100, "Size of each value");
+DEFINE_string(distribution_type, "fixed",
```
Comment (zhichao-cao) [inline tools/db_bench_tool.cc]: `value_size_distribution_type` might be more clear for users.

File: tools/db_bench_tool.cc [hunk 2]
```diff
@@ -1416,20 +1429,126 @@ class ReportFileOpEnv : public EnvWrapper {
...
+      val = std::min(max_value_size_, val);
+    }
+    return val;
+  }
+ private:
+  virtual unsigned int Get(std::mt19937&) = 0;
+  virtual bool NeedTruncate() {
+    return false;
+  }
+  unsigned int min_value_size_;
```
Comment (zhichao-cao) [inline tools/db_bench_tool.cc]: Maybe just provide the interface in the BaseDistribution. Other distributions like Zipfian or Exp might can also inherited from this class. In that case, the private member such as  std::random_device, std::mt19937 cannot be used.

File: tools/db_bench_tool.cc [hunk 3]
```diff
@@ -1447,13 +1566,9 @@ class RandomGenerator {
     return Slice(data_.data() + pos_ - len, len);
   }
 
-  Slice GenerateWithTTL(unsigned int len) {
```
Comment (zhichao-cao) [inline tools/db_bench_tool.cc]: Why this function is removed?

Comment (burtonli) [reply tools/db_bench_tool.cc]: 1. GenerateWithTTL is dead code, no one calls it.
2. GenerateWithTTL does the same with thing as 'Slice Generate(unsigned int len)'

File: tools/db_bench_tool.cc (hunk 1 above)
Comment (burtonli) [reply tools/db_bench_tool.cc]: Thanks, I will update

File: tools/db_bench_tool.cc [hunk 4]
```diff
@@ -1416,20 +1429,129 @@ class ReportFileOpEnv : public EnvWrapper {
...
 class RandomGenerator {
  private:
   std::string data_;
   unsigned int pos_;
+  std::unique_ptr<BaseDistribution> dist_;
 
  public:
+
   RandomGenerator() {
+    auto max_value_size = FLAGS_value_size_max;
```
Comment (zhichao-cao) [inline tools/db_bench_tool.cc]: Using switch can make it easier to extend to other distributions

File: tools/db_bench_tool.cc [hunk 5]
```diff
@@ -1416,20 +1429,129 @@ class ReportFileOpEnv : public EnvWrapper {
...
+    if (NeedTruncate()) {
+      val = std::max(min_value_size_, val);
+      val = std::min(max_value_size_, val);
+    }
+    return val;
+  }
+ private:
+  virtual unsigned int Get() = 0;
+  virtual bool NeedTruncate() {
+    return false;
```
Comment (zhichao-cao) [inline tools/db_bench_tool.cc]: It is safer to enable value size truncate in BaseDistribution. I suggest to set it to "true".

File: tools/db_bench_tool.cc [hunk 6]
```diff
@@ -5343,24 +5472,24 @@ class Benchmark {
       } else if (query_type == 1) {
         // the Put query
         puts++;
-        int64_t value_size = ParetoCdfInversion(
+        int64_t size = ParetoCdfInversion(
```
Comment (zhichao-cao) [inline tools/db_bench_tool.cc]: `size` can be misleading. Maybe v_size or others.

File: tools/db_bench_tool.cc [hunk 7]
```diff
@@ -6106,21 +6237,22 @@ class Benchmark {
       GenerateKeyFromInt(key_rand, merge_keys_, &key);
 
       Status s;
+      auto val = gen.Generate();
```
Comment (zhichao-cao) [inline tools/db_bench_tool.cc]: I suggest not to use `auto` here for val. The same in other places.

**Comment by zhichao-cao** [comment]
LGTM, thanks for the contribution.

**Pull request closed without merging**